image/mkosi/upload/upload_azure.sh

```diff
@@ -0,0 +1,131 @@
+elif [[ -z "${AZURE_VMGS_PATH}" ]] && [[ "${AZURE_SECURITY_TYPE}" == "TrustedLaunch"  ]]; then
+    echo "No VMGS path provided - using default TrsutedLaunch VMGS"
+    AZURE_VMGS_PATH="${BLOBS_DIR}/trusted-launch-vmgs.vhd"
+fi
+
+SIZE=$(wc -c "${AZURE_IMAGE_PATH}" | cut -d " " -f1)
+
+create_disk () {
+    az disk create \
+        -n "${AZURE_DISK_NAME}" \
+        -g "${AZURE_RESOURCE_GROUP_NAME}" \
+        -l "${AZURE_REGION}" \
+        --hyper-v-generation V2 \
+        --os-type Linux \
+        --upload-size-bytes "${SIZE}" \
+        --sku standard_lrs \
+        --upload-type UploadWithSecurityData \
+        --security-type "${AZURE_DISK_SECURITY_TYPE}"
+    az disk wait --created -n "${AZURE_DISK_NAME}" -g "${AZURE_RESOURCE_GROUP_NAME}"
+    az disk list --output table --query "[?name == '${AZURE_DISK_NAME}' && resourceGroup == '${AZURE_RESOURCE_GROUP_NAME^^}']"
+    DISK_SAS=$(az disk grant-access -n ${AZURE_DISK_NAME} -g ${AZURE_RESOURCE_GROUP_NAME} \
+        --access-level Write --duration-in-seconds 86400 \
+         ${AZURE_VMGS_PATH+"--secure-vm-guest-state-sas"})
+    azcopy copy "${AZURE_IMAGE_PATH}" \
+        "$(echo $DISK_SAS | jq -r .accessSas)" \
+        --blob-type PageBlob
+    if [[ -z "${AZURE_VMGS_PATH}" ]]; then
+        echo "No VMGS path provided - skipping VMGS upload"
+    else
+        azcopy copy "${AZURE_VMGS_PATH}" \
+            "$(echo $DISK_SAS | jq -r .securityDataAccessSas)" \
+            --blob-type PageBlob
+    fi
+    az disk revoke-access -n "${AZURE_DISK_NAME}" -g "${AZURE_RESOURCE_GROUP_NAME}"
+}
+
+delete_disk () {
+    az disk delete -y -n "${AZURE_DISK_NAME}" -g "${AZURE_RESOURCE_GROUP_NAME}"
+}
+
+create_sig_version () {
+    az sig create -l "${AZURE_REGION}" --gallery-name "${AZURE_GALLERY_NAME}" --resource-group "${AZURE_RESOURCE_GROUP_NAME}" || true
+    az sig image-definition create \
+        --resource-group "${AZURE_RESOURCE_GROUP_NAME}" \
+        -l "${AZURE_REGION}" \
+        --gallery-name "${AZURE_GALLERY_NAME}" \
+        --gallery-image-definition "${AZURE_IMAGE_DEFINITION}" \
+        --publisher "${AZURE_PUBLISHER}" \
+        --offer "${AZURE_IMAGE_OFFER}" \
+        --sku "${AZURE_SKU}" \
+        --os-type Linux \
+        --os-state generalized \
+        --hyper-v-generation V2 \
+        --features SecurityType="${AZURE_SECURITY_TYPE}" || true
+    az sig image-version create \
+        --resource-group "${AZURE_RESOURCE_GROUP_NAME}" \
+        -l "${AZURE_REGION}" \
+        --gallery-name "${AZURE_GALLERY_NAME}" \
+        --gallery-image-definition "${AZURE_IMAGE_DEFINITION}" \
+        --gallery-image-version "${AZURE_IMAGE_VERSION}" \
+        --target-regions ${AZURE_REPLICATION_REGIONS} \
+        ${AZURE_CVM_ENCRYPTION_ARGS} \
+        --replica-count 1 \
+        --replication-mode Full \
+        --os-snapshot "$(az disk list --query "[?name == '${AZURE_DISK_NAME}' && resourceGroup == '${AZURE_RESOURCE_GROUP_NAME^^}'] | [0].id" --output tsv)"
```

Comment:
(Whole file)
Personal opinion but for some more complex scripts I would maybe suggesting switching to Python or similar instead of bash for easier maintenance and readability.

Reply:
I'll try to refactor this a bit.

Reply:
I tried to simplify this a bit. Rewriting this is python is not a good solution IMO as this script mostly calls `az` with different arguments.